adapters/dkb-adapter/src/main/java/de/adorsys/xs2a/gateway/service/impl/DkbAccessTokenService.java

```diff
@@ -0,0 +1,150 @@
+    }
+
+    static {
+        String consumerKey = System.getProperty(DKB_TOKEN_CONSUMER_KEY_PROPERTY);
+        String consumerSecret = System.getProperty(DKB_TOKEN_CONSUMER_SECRET_PROPERTY);
+
+        if (consumerKey.isEmpty() || consumerSecret.isEmpty()) {
+            throw new AccessTokenException("Consumer key or secret are not provided");
+        }
+
+        headers = new HashMap<>();
+        headers.put("Content-type", "application/x-www-form-urlencoded");
+        headers.put("Authorization", "Basic " + buildBasicAuthorization(consumerKey, consumerSecret));
+
+        tokenUrl = System.getProperty(DKB_TOKEN_URL_PROPERTY, DEFAULT_TOKEN_URL);
+
+        secondsBeforeTokenExpiration = Integer.valueOf(System.getProperty(
+                DKB_TOKEN_SECONDS_BEFORE_TOKEN_EXPIRATION_PROPERTY,
+                DEFAULT_SECONDS_BEFORE_TOKEN_EXPIRATION
+        ));
+    }
+
+    @Override
+    public String retrieveToken() {
+        if (!isValid()) {
+            GeneralResponse<TokenResponse> response = httpClient.post(
+                    tokenUrl,
+                    "grant_type=client_credentials",
+                    headers,
+                    responseHandler()
+            );
+            accessToken = new AccessToken(response.getResponseBody());
+        }
+        return accessToken.token;
+    }
+
+    void setHttpClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    private boolean isValid() {
+        if (accessToken == null || accessToken.token == null || accessToken.validity == null) {
+            return false;
+        }
+        LocalDateTime expirationDate = LocalDateTime.now().plusSeconds(secondsBeforeTokenExpiration);
+        return accessToken.validity.isAfter(expirationDate);
+    }
+
+    private static String buildBasicAuthorization(String key, String secret) {
+        String credentials = key + ":" + secret;
+        return new String(Base64.getEncoder().encode(credentials.getBytes()));
+    }
+
+    private HttpClient.ResponseHandler<TokenResponse> responseHandler() {
+        return (statusCode, responseBody, responseHeaders) -> {
+            if (statusCode == 200) {
+                return jsonMapper.readValue(responseBody, TokenResponse.class);
+            }
+            throw new AccessTokenException("Can't retrieve access token by provided credentials");
+        };
+    }
+
+    static class TokenResponse {
+        // {"access_token":"9c60cc0a-7b38-3a2a-9b43-2b7d9d957838","scope":"am_application_scope default","token_type":"Bearer","expires_in":3600}
+        @JsonProperty("access_token")
+        private String accessToken;
+        @JsonProperty("expires_in")
+        private int expiresIn;
+
+        TokenResponse() {
+        }
+
+        TokenResponse(String accessToken, int expiresIn) {
+            this.accessToken = accessToken;
+            this.expiresIn = expiresIn;
+        }
+    }
+
+    private static class AccessToken {
+        private String token = "";
```

Comment:
The initialization of this variable is redundant.

Reply:
Thanks, removed